README.md
# flutter_annual_task

## flutter_annual_task
Flutter package for displaying grid view of daily task like **Github-Contributions**.

**Example**
![example](https://github.com/HuanSuh/flutter_annual_task/blob/master/example/assets/example.png?raw=true)

## Usage

Make sure to check out [example project](https://github.com/HuanSuh/flutter_annual_task/tree/master/example).

```dart
AnnualTaskView(
  taskItem	// List<AnnualTaskItem>
),
```

- - -
### Installation
Add to pubspec.yaml:
```xml
dependencies:
  flutter_annual_task: ^0.1.3
```
Then import it to your project:
```dart
import 'package:flutter_annual_task/flutter_annual_task.dart';
```
And finally add `AnnualTaskView` widget in your project.
```dart
AnnualTaskView(
  taskItem	// List<AnnualTaskItem>
),
```
- - -

### AnnualTaskItem
**AnnualTaskItem**

```dart
class AnnualTaskItem {
  final DateTime date;
  final double proceeding;	// 0.0 ~ 1.0

  AnnualTaskItem(this.date, [this.proceeding = 1.0]);
  ...
}
```
The value of `proceeding` affects the opacity on the each cell of daily task.<br/>- For showing the color in visual, the minimum value of displaying is 80(max: 255).
</br>

**AnnualTaskColorItem**

If you want to specify color for each daily task, you can use `AnnualTaskColorItem`.
```dart
class AnnualTaskColorItem extends AnnualTaskItem {
  final Color color;

  AnnualTaskColorItem(
    DateTime date, {
    double proceeding = 1.0,
    this.color,
  }) : super(date, proceeding);
  ...
}
```

You should generate **list of AnnualTaskItem**(`List<AnnualTaskItem>`) to use this package.</br>Below is an example for building  **list of AnnualTaskItem**.
```dart
//AnnualTaskItem
<your_item_list>.map(
  (item) => AnnualTaskItem(
    item.date,
    0.5,
  ),
)
.toList();

//AnnualTaskColorItem
<your_item_list>.map(
  (item) => AnnualTaskColorItem(
    item.date,
    color: <color_for_each_item>
  ),
)
.toList();
```
## Examples
### Cell Shape
Specify cellShape with `AnnualTaskCellShape` with `AnnualTaskCellShape.ROUNDED_SQUARE`(default), `AnnualTaskCellShape.SQUARE` or `AnnualTaskCellShape.CIRCLE`.

**Square**

| ![Image](https://github.com/HuanSuh/flutter_annual_task/blob/master/example/assets/example_cellshape_square.png?raw=true) |
| :---: |
| square |
```dart
AnnualTaskView(
  taskItem, // List<AnnualTaskItem>
  cellShape: AnnualTaskCellShape.SQUARE,
)
```
**Circle**

| ![Image](https://github.com/HuanSuh/flutter_annual_task/blob/master/example/assets/example_cellshape_circle.png?raw=true) |
| :---: |
| circle |

```dart
AnnualTaskView(
  taskItem, // List<AnnualTaskItem>
  cellShape: AnnualTaskCellShape.CIRCLE,
)
```

**AnnualTaskColorItem**

| ![Image](https://github.com/HuanSuh/flutter_annual_task/blob/master/example/assets/example_cellshape_coloritem.png?raw=true) |
| :---: |
| circle |

```dart
AnnualTaskView(
  taskItemWithColor, // List<AnnualTaskColorItem>
)
```

### Labels
You can edit the **labels of week** or the **labels of month**.
```dart
AnnualTaskView(
  taskItem, // List<AnnualTaskItem>
  showMonthLabel: false, //default : true
  showWeekDayLabel: false, //default : true
)
```

| ![Image](https://github.com/HuanSuh/flutter_annual_task/blob/master/example/assets/example_label_without.png?raw=true) |
| :---: |
| without labels |

**Custom label**

| ![Image](https://github.com/HuanSuh/flutter_annual_task/blob/master/example/assets/example_label_custom.png?raw=true) |
| :---: |
| custom labels |
```dart
AnnualTaskView(
  taskItem, // List<AnnualTaskItem>
  weekDayLabels: ['', 'Mon', '', 'Wed', '', 'Fri', ''],
  monthLabels: ['1','2','3','4','5','6','7','8','9','10','11','12'],
)
```
The type of `weekDayLabels` and `monthLabels` is `List<String>`.

- `weekDayLabels` starts from Sunday.
- default value of `weekDayLabels' is ['S', 'M', 'T', 'W', 'T', 'F', 'S'].
- default value of `monthLabels' is `['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']`.
- You can also hide the label of each items with empty String(`''`). But, `weekDayLabels` should be **length of 7** and, `monthLabels` should be **length of 12**.

**Styled label**

| ![Image](https://github.com/HuanSuh/flutter_annual_task/blob/master/example/assets/example_label_style.png?raw=true) |
| :---: |
| Styled label |
```dart
AnnualTaskView(
  taskItem, // List<AnnualTaskItem>
  labelStyle: TextStyle(
    color: Colors.blueGrey,
    fontSize: 10,
    fontWeight: FontWeight.bold,
  ),	// default label style : TextStyle(fontSize: 8)
)
```
## Props

| props | type | desc |
| :--- | :--- | :--------------- |
| items | `List<AnnualTaskItem>` | List of `AnnualTaskItem` |
| year | `int` | default : `DateTime.now().year` |
| activateColor | `Color` | default : `Theme.of(context).primaryColor` |
| emptyColor | `Color` | Color of cell with proceeding `0.0` or the day which `items` doesn't contain. <br/>default : `Color(0xFFD0D0D0)` |
| showWeekDayLabel | `bool` | Show the labels of week, if true.<br/>default : `true` |
| cellShape | `AnnualTaskCellShape` | Shape of cell. One of `AnnualTaskCellShape.ROUNDED_SQUARE`, `AnnualTaskCellShape.SQUARE` or `AnnualTaskCellShape.CIRCLE`.<br/>default: `AnnualTaskCellShape.ROUNDED_SQUARE` |
| showMonthLabel | `bool` | Show the labels of month, if true.<br/>default : `true` |
| monthLabels | `List<String>` | Labels of month.<br/>default: `['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']` |
| weekDayLabels | `List<String>` | Labels of week.<br/>default: `['S', 'M', 'T', 'W', 'T', 'F', 'S']` |
| labelStyle | `TextStyle` | `TextStyle` of labels.<br/>default: `TextStyle(fontSize: 8)` |


CHANGELOG.md
## [0.2.0] - Mar 7, 2021
* null-safety

## [0.1.4] - Mar 8, 2020
* minor bug fix
   * Set week label align to Alignment.centerRight.

## [0.1.3] - Mar 2, 2020
* minor bug fix.

## [0.1.2] - Mar 1, 2020
* Update package information.

## [0.1.0] - Mar 1, 2020
* Draft


lib/src/annual_task_view.dart
import 'dart:async';

import 'package:flutter/material.dart';
import './annual_task_item.dart';

enum AnnualTaskCellShape { SQUARE, CIRCLE, ROUNDED_SQUARE }

class AnnualTaskView extends StatefulWidget {
  final List<AnnualTaskItem> items;
  final Color? activateColor;
  final Color? emptyColor;
  final int year;

  final bool showWeekDayLabel;
  final List<String>? weekDayLabels;
  final bool showMonthLabel;
  final List<String>? monthLabels;
  final AnnualTaskCellShape? cellShape;

  final bool swipeEnabled;
  final TextStyle? labelStyle;

  AnnualTaskView(
    this.items, {
    int? year,
    this.activateColor,
    this.emptyColor,
    this.cellShape,
    this.showWeekDayLabel = true,
    this.showMonthLabel = true,
    List<String>? weekDayLabels,
    List<String>? monthLabels,
    this.labelStyle,
    this.swipeEnabled = true,
  })  : assert(showWeekDayLabel == false ||
            (weekDayLabels == null || weekDayLabels.length == 7)),
        assert(showMonthLabel == false ||
            (monthLabels == null || monthLabels.length == 12)),
        this.year = year ?? DateTime.now().year,
        this.weekDayLabels =
            (weekDayLabels?.length ?? 0) == 7 ? weekDayLabels : _WEEKDAY_LABELS,
        this.monthLabels =
            (monthLabels?.length ?? 0) == 12 ? monthLabels : _MONTH_LABELS;

  @override
  _AnnualTaskViewState createState() => _AnnualTaskViewState();
}

class _AnnualTaskViewState extends State<AnnualTaskView> {
  double? contentsWidth;
  bool? _refreshing;

  final StreamController<Map<DateTime, AnnualTaskItem>?> _streamController =
      StreamController();

  @override
  void initState() {
    super.initState();
    Future.delayed(Duration.zero, () => _buildListToMap());
  }

  @override
  void didUpdateWidget(AnnualTaskView oldWidget) {
    super.didUpdateWidget(oldWidget);
    _buildListToMap();
  }

  @override
  void dispose() {
    _streamController.close();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        _layoutManagerView(),
        StreamBuilder<Map<DateTime, AnnualTaskItem>?>(
          stream: _streamController.stream,
          builder: (context, snapshot) {
            double opacity = 1.0;
            if (contentsWidth == null) {
              opacity = 0.0;
            } else if (_refreshing == true) {
              opacity = 0.5;
            }
            return AnimatedOpacity(
              opacity: opacity,
              duration: Duration(milliseconds: opacity == 1.0 ? 500 : 0),
              child: _AnnualTaskGrid(
                widget.year,
                snapshot.data,
                widget.activateColor ?? Theme.of(context).primaryColor,
                widget.emptyColor,
                widget.showWeekDayLabel,
                widget.showMonthLabel,
                widget.weekDayLabels,
                widget.monthLabels,
                widget.cellShape,
                widget.labelStyle,
                contentsWidth,
              ),
            );
          },
        ),
      ],
    );
  }

  Future<Map<DateTime, AnnualTaskItem>> _buildListToMap() {
    Map<DateTime, AnnualTaskItem> resultMap = Map();
    _refreshing = true;
    _streamController.add(null);
    widget.items.forEach((item) {
      resultMap[item.date] = item;
    });
    return Future.delayed(Duration.zero, () {
      _streamController.add(resultMap);
      _refreshing = false;
      return resultMap;
    });
  }

  Widget _layoutManagerView() {
    if (contentsWidth != null) return Container();
    return Container(
      height: 0,
      child: Row(
        children: <Widget>[
          (widget.showWeekDayLabel != false)
              ? Column(
                  children: List.generate(7, (idx) {
                  return Text(
                    widget.weekDayLabels?.elementAt(idx) ?? '',
                    style: widget.labelStyle ?? _LABEL_STYLE,
                  );
                }))
              : Container(),
          Expanded(
            child: LayoutBuilder(builder: (_, layout) {
              Future.delayed(
                Duration.zero,
                () => setState(() => contentsWidth = layout.maxWidth),
              );
              return Container();
            }),
          )
        ],
      ),
    );
  }
}

class _AnnualTaskGrid extends StatelessWidget {
  final DateTime firstDate;
  final Map<DateTime, AnnualTaskItem>? resultMap;
  final int firstDay;

  final Color? activateColor;
  final Color? emptyColor;

  final bool showWeekDayLabel;
  final List<String>? weekDayLabels;
  final bool showMonthLabel;
  final List<String>? monthLabels;
  final AnnualTaskCellShape cellShape;
  final TextStyle labelStyle;

  final double? contentsWidth;

  _AnnualTaskGrid(
    int year,
    this.resultMap,
    this.activateColor,
    this.emptyColor,
    this.showWeekDayLabel,
    this.showMonthLabel,
    this.weekDayLabels,
    this.monthLabels,
    AnnualTaskCellShape? cellShape,
    TextStyle? labelStyle,
    this.contentsWidth,
  )   : firstDate = DateTime(year, 1, 1),
        firstDay = DateTime(year, 1, 1).weekday % 7,
        this.cellShape = cellShape ?? AnnualTaskCellShape.ROUNDED_SQUARE,
        this.labelStyle = labelStyle ?? _LABEL_STYLE;

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, layout) {
        double maxWidth = contentsWidth ?? layout.maxWidth;
        final double cellSize = (maxWidth / 53) * 0.85;
        return Column(
          children: List.generate(
            _rowCnt,
            (days) {
              if (showMonthLabel == true && days == 0) {
                return _buildMonthLabelRow(
                  cellSize,
                  paddingLeft: layout.maxWidth - maxWidth,
                );
              }
              return Padding(
                padding:
                    EdgeInsets.symmetric(vertical: (maxWidth / 53) * 0.075),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceAround,
                  children: List.generate(_colCnt, (weeks) {
                    if (showWeekDayLabel == true && weeks == 0) {
                      return _buildWeekdayLabel(days,
                          width: layout.maxWidth - maxWidth);
                    }
                    AnnualTaskItem? result = _getResult(weeks, days);
                    return ClipRRect(
                      borderRadius: BorderRadius.circular(
                        cellShape == AnnualTaskCellShape.SQUARE
                            ? 0
                            : cellShape == AnnualTaskCellShape.CIRCLE
                                ? 200
                                : cellSize / 4,
                      ),
                      child: Container(
                        width: cellSize,
                        height: cellSize,
                        color: result?.fillColor(activateColor ??
                                Theme.of(context).accentColor) ??
                            (emptyColor ?? Theme.of(context).disabledColor),
                      ),
                    );
                  }),
                ),
              );
            },
          ),
        );
      },
    );
  }

  AnnualTaskItem? _getResult(int weeksIdx, int daysIdx) {
    if (resultMap == null) return null;
    if (showWeekDayLabel) weeksIdx--;
    if (showMonthLabel) daysIdx--;
    int days = weeksIdx * 7 + daysIdx - firstDay;
    DateTime date = firstDate.add(Duration(days: days));
    return resultMap![date];
  }

  int get _colCnt => showWeekDayLabel == true ? 54 : 53;
  int get _rowCnt => showMonthLabel == true ? 8 : 7;

  Widget _buildMonthLabelRow(double cellSize, {double? paddingLeft}) {
    return Padding(
      padding: EdgeInsets.only(left: paddingLeft ?? 0),
      child: Row(
        children: List.generate(12, (idx) {
          return _buildMonthLabel(idx, cellSize);
        }),
      ),
    );
  }

  Widget _buildMonthLabel(int idx, double cellSize) {
    DateTime date =
        DateTime(firstDate.year, idx + 2, 1).subtract(Duration(days: 1));
    return Container(
      width: (cellSize / 0.85) * (date.day / 7.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Container(
            constraints: BoxConstraints(minWidth: cellSize * 1.5),
            child: Text(
              monthLabels?.elementAt(idx) ?? '',
              style: labelStyle,
              textAlign: TextAlign.center,
            ),
          )
        ],
      ),
    );
  }

  Widget _buildWeekdayLabel(int weekIdx, {double? width}) {
    return Container(
      width: width ?? 0,
      alignment: Alignment.centerRight,
      child: Text(
        weekDayLabels?.elementAt(weekIdx - (showMonthLabel ? 1 : 0)) ?? '',
        style: labelStyle,
      ),
    );
  }
}

const List<String> _WEEKDAY_LABELS = ['S', 'M', 'T', 'W', 'T', 'F', 'S'];
const List<String> _MONTH_LABELS = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec'
];
const TextStyle _LABEL_STYLE = TextStyle(fontSize: 8);


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_annual_task/flutter_annual_task.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    List<AnnualTaskItem> taskItems = AnnualTaskItemHelper.generateAnnualTask();
    List<AnnualTaskItem> taskItemsWithColor =
        AnnualTaskItemHelper.generateAnnualTaskColorItem(taskItems);
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('annual_task_view'),
        ),
        body: RefreshIndicator(
          onRefresh: () {
            setState(() {});
            return Future.delayed(Duration(milliseconds: 500));
          },
          child: SingleChildScrollView(
            child: Padding(
              padding: const EdgeInsets.all(8.0),
              child: Column(
                children: [
                  _buildSectionTitle('Default'),
                  _buildExample(
                    '',
                    AnnualTaskView(
                      taskItems,
                      showMonthLabel: false,
                      showWeekDayLabel: false,
                    ),
                  ),
                  // Examples for cellShape
                  _buildSectionTitle('Cell shape'),
                  _buildExample(
                    'square',
                    AnnualTaskView(
                      taskItems, // List<AnnualTaskItem>
                      activateColor: Colors.red,
                      cellShape: AnnualTaskCellShape.SQUARE,
                    ),
                  ),
                  _buildExample(
                    'circle',
                    AnnualTaskView(
                      taskItems, // List<AnnualTaskItem>
                      activateColor: Colors.red,
                      emptyColor: Colors.grey.withAlpha(80),
                      cellShape: AnnualTaskCellShape.CIRCLE,
                    ),
                  ),
                  _buildExample(
                    'AnnualTaskColorItem',
                    AnnualTaskView(
                      taskItemsWithColor,
                      weekDayLabels: [
                        'Sun',
                        'Mon',
                        'Tue',
                        'Wed',
                        'Thu',
                        'Fri',
                        'Sat'
                      ],
                    ),
                  ),

                  // Examples for label
                  _buildSectionTitle('Labels'),
                  _buildExample(
                    'without labels',
                    AnnualTaskView(
                      taskItems, // List<AnnualTaskItem>
                      showMonthLabel: false,
                      showWeekDayLabel: false,
                    ),
                  ),
                  _buildExample(
                    'custom labels',
                    AnnualTaskView(
                      taskItems, // List<AnnualTaskItem>
                      weekDayLabels: CUSTOM_WEEKDAY_LABEL,
                      monthLabels: CUSTOM_MONTH_LABEL,
                    ),
                  ),
                  _buildExample(
                    'labels style',
                    AnnualTaskView(
                      taskItems, // List<AnnualTaskItem>
                      weekDayLabels: CUSTOM_WEEKDAY_LABEL,
                      monthLabels: CUSTOM_MONTH_LABEL,
                      labelStyle: TextStyle(
                        color: Colors.blueGrey,
                        fontSize: 10,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }

  Widget _buildSectionTitle(String title) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 8.0),
      child: Text(title, style: Theme.of(context).textTheme.headline6),
    );
  }

  Widget _buildExample(String title, AnnualTaskView child) {
    return Column(children: [
      title?.isNotEmpty == true ? Text(title) : Container(),
      SizedBox(height: 4.0),
      child,
      Divider(),
    ]);
  }
}

const List<String> CUSTOM_WEEKDAY_LABEL = ['', 'Mon', '', 'Wed', '', 'Fri', ''];
const List<String> CUSTOM_MONTH_LABEL = [
  'Jan',
  '',
  'Mar',
  '',
  'May',
  '',
  'Jul',
  '',
  'Sep',
  '',
  'Nov',
  ''
];

class AnnualTaskItemHelper {
  static List<AnnualTaskItem> generateAnnualTask({int year, int sampleSize}) {
    var rnd = new Random();
    sampleSize = sampleSize ?? max(80, min(365, rnd.nextInt(200)));
    year ??= DateTime.now().year;
    DateTime prevDate = DateTime(year, 1, 1);
    return List.generate(sampleSize, (idx) {
      int maxDiff =
          (365 - prevDate.difference(DateTime(year, 12, 31)).inDays) ~/
              (sampleSize - idx);
      prevDate = prevDate.add(Duration(days: rnd.nextInt(maxDiff) + 1));
      return AnnualTaskItem(prevDate, rnd.nextDouble());
    });
  }

  static List<AnnualTaskColorItem> generateAnnualTaskColorItem(
      List<AnnualTaskItem> items) {
    List<Color> colors = [Colors.red, Colors.blue, Colors.orange];
    return items
        .map((e) => AnnualTaskColorItem(
              e.date,
              proceeding: e.proceeding,
              color: colors[Random().nextInt(colors.length)],
            ))
        .toList();
  }
}
